### PR TITLE
statistics, ranger: add method to calculate the range and row count of non pk column

### PR DESCRIPTION
**Pull request by winoros**
Calculate the range of non pk column.
Calculate the row count by non pk column histogram.

PTAL @shenli @hanfei1991 @XuHuaiyu @lamxTyler @coocood

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -338,6 +338,53 @@ func (c *Column) getIntColumnRowCount(sc *variable.StatementContext, intRanges [
 	return rowCount, nil
 }
 
+func (c *Column) getColumnRowCount(sc *variable.StatementContext, ranges ...types.ColumnRange) (float64, error) {
+	var rowCount float64
+	for _, rg := range ranges {
+		// the point case.
```
Comment (hanfei1991) [inline statistics/histogram.go]: Put this comment to line 350

File: util/ranger/refiner.go [hunk 2]
```diff
@@ -300,6 +300,27 @@ func BuildTableRange(accessConditions []expression.Expression, sc *variable.Stat
 	return ranges, nil
 }
 
+// BuildColumnRange builds the range for sampling histogram to calculate the row count.
+func BuildColumnRange(conds []expression.Expression, sc *variable.StatementContext, tp *types.FieldType) ([]*types.ColumnRange, error) {
```
Comment (hanfei1991) [inline util/ranger/refiner.go]: Should we figure out a particular column to build range?

File: util/ranger/range.go [hunk 3]
```diff
@@ -626,3 +626,27 @@ func (r *Builder) BuildTableRanges(rangePoints []point) []types.IntColumnRange {
 	}
 	return tableRanges
 }
+
+func (r *Builder) buildColumnRanges(points []point, tp *types.FieldType) []*types.ColumnRange {
+	columnRanges := make([]*types.ColumnRange, 0, len(points)/2)
+	for i := 0; i < len(points); i += 2 {
+		startPoint := r.convertPoint(points[i], tp)
+		endPoint := r.convertPoint(points[i+1], tp)
```
Comment (lamxTyler) [inline util/ranger/range.go]: Is it always guaranteed that the `len(points)` is an even number? If not, `i+1` may equal to `len(points)`.

File: util/ranger/refiner.go [hunk 4]
```diff
@@ -251,15 +251,15 @@ func DetachIndexScanConditions(conditions []expression.Expression, index *model.
 	return accessConds, filterConds, accessEqualCount, accessInAndEqCount
 }
 
-// DetachTableScanConditions distinguishes between access conditions and filter conditions from conditions.
-func DetachTableScanConditions(conditions []expression.Expression, pkName model.CIStr) ([]expression.Expression, []expression.Expression) {
-	if pkName.L == "" {
+// DetachColumnConditions distinguishes between access conditions and filter conditions from conditions.
+func DetachColumnConditions(conditions []expression.Expression, colName model.CIStr) ([]expression.Expression, []expression.Expression) {
```
Comment (lamxTyler) [inline util/ranger/refiner.go]: Why rename it to `colName`?

**Comment by hanfei1991** [comment]
It's better to use a more common function:
buildRangesByColumns(conditions []Expression, sc, columns []*Columns) (accessionConds, filterConds, ranges, err)

This function can process both index columns and single column. **And you can only use it for new plan.**

**Comment by winoros** [comment]
@hanfei1991 This function will added in a new pr. PTAL
@shenli @coocood PTAL

File: util/ranger/refiner.go (hunk 4 above)
Comment (winoros) [inline util/ranger/refiner.go]: This function will process cases that the column isn't a pk.

File: util/ranger/range.go (hunk 3 above)
Comment (winoros) [inline util/ranger/range.go]: Yes, it's guaranteed.

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@shenli PTAL

File: statistics/histogram.go [hunk 5]
```diff
@@ -376,6 +376,53 @@ func (c *Column) getIntColumnRowCount(sc *variable.StatementContext, intRanges [
...
+			if err != nil {
+				return 0, errors.Trace(err)
+			}
+			cnt += highCnt
+		}
+		rowCount += cnt
+	}
+	if rowCount > c.totalRowCount() {
+		rowCount = c.totalRowCount()
+	} else if rowCount < 0 {
```
Comment (shenli) [inline statistics/histogram.go]: Is it possible that rowCount < 0?

File: statistics/histogram.go [hunk 6]
```diff
@@ -376,6 +376,53 @@ func (c *Column) getIntColumnRowCount(sc *variable.StatementContext, intRanges [
 	return rowCount, nil
 }
 
+func (c *Column) getColumnRowCount(sc *variable.StatementContext, ranges ...types.ColumnRange) (float64, error) {
```
Comment (shenli) [inline statistics/histogram.go]: Add comment for the function.

File: util/ranger/refiner.go [hunk 7]
```diff
@@ -251,15 +251,15 @@ func DetachIndexScanConditions(conditions []expression.Expression, index *model.
...
+	if colName.L == "" {
 		return nil, conditions
 	}
 
 	var accessConditions, filterConditions []expression.Expression
 	checker := conditionChecker{
-		pkName: pkName,
+		pkName: colName,
```
Comment (shenli) [inline util/ranger/refiner.go]: Should we rename pkName?

File: statistics/histogram.go (hunk 5 above)
Comment (winoros) [inline statistics/histogram.go]: Since the calculation is approximate, there may be this case.

File: util/ranger/refiner.go (hunk 7 above)
Comment (winoros) [inline util/ranger/refiner.go]: Yes.

**Comment by shenli** [comment]
LGTM

**Pull request merged**